File: java/src/jan/week2/boj/커피숍2_박재환.java
package jan.week2.boj;

import java.util.*;
import java.io.*;

public class 커피숍2_박재환 {
    static BufferedReader br;
    static StringBuilder sb;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        init();
        br.close();
        System.out.println(sb);
    }

    /**+
     *  연속으로 구간합을 구해야한다.
     *  - DP로 모든 구간 합을 저장하는 방법
     *  - 세그먼트 트리
     *
     *  단, 문제에서 'N번재 수를 고쳐야한다' 는 조건이 있다.
     *  -> 반복적인 중간 값 교체가 일어난다.
     *
     *  => Segment Tree or Lazy Propagation Segment Tree
     *
     *  => 조회, 교체가 한 번씩 일어남 => 그냥 Segment Tree 로 가능
     */
    static StringTokenizer st;
    static int n, q;
    static int[] arr;
    static long[] tree;
    static void init() throws IOException {
        st = new StringTokenizer(br.readLine().trim());
        n = Integer.parseInt(st.nextToken());       // 정수의 개수
        q = Integer.parseInt(st.nextToken());       // 턴의 수
        arr = new int[n];
        st = new StringTokenizer(br.readLine().trim());
        for(int i=0; i<n; i++) arr[i] = Integer.parseInt(st.nextToken());

        // 1. Segment Tree 생성
        tree = new long[4*n];
        makeTree(0, n-1, 1);

        while(q-- > 0) {
            st = new StringTokenizer(br.readLine().trim());
            // Note : x > y 로 주어지는 경우, y~x 번째이다.
            int l = Integer.parseInt(st.nextToken())-1;
            int r = Integer.parseInt(st.nextToken())-1;
            if(r < l) {
                int tmp = r;
                r = l;
                l = tmp;
            }
            sb.append(query(0, n-1, 1, l, r)).append('\n');
            int targetId = Integer.parseInt(st.nextToken())-1;
            int targetValue = Integer.parseInt(st.nextToken());
            update(0, n-1, 1, targetId, targetValue);
            arr[targetId] = targetValue;
        }
    }

    static long makeTree(int s, int e, int id) {
        if(s == e) return tree[id] = arr[s];        // 리프 노드에 도달한 경우
        // 하위 노드부터 탐색해서 바텀업으로 값을 채움
        int mid = s + (e-s)/2;
        return tree[id] = makeTree(s, mid, 2*id) + makeTree(mid+1, e, 2*id+1);
    }

    static Long query(int s, int e, int id, int l, int r) {
        if(s > r || e < l) return 0L;           // query의 범위를 완전하게 벗어나는 경우

        if(s >= l && e <= r) return tree[id];   // query 가 범위를 완전하게 포함하는 경우
        // 더 나눠서 탐색해야하는 경우
        int mid = s + (e-s)/2;
        return query(s, mid, 2*id, l, r) + query(mid+1, e, 2*id+1, l, r);
    }

    static void update(int s, int e, int id, int targetId, int targetValue) {
        if(targetId < s || targetId > e) return;    // update 범위를 완전히 벗어나는 경우

        // 값을 교체하는 것
        tree[id] = tree[id] + targetValue - arr[targetId];
        if(s == e) return;
        int mid = s + (e-s)/2;
        update(s, mid, 2*id, targetId, targetValue);
        update(mid+1, e, 2*id+1, targetId, targetValue);
    }
}

File: python/src/jan/week2/boj/커피숍2_박재환.py
import sys
input = sys.stdin.readline

n, q = map(int, input().split())
arr = list(map(int, input().split()))
tree = [0] * (4*n)

def make_tree(s, e, id):
    if s == e:
        tree[id] = arr[s]
        return tree[id]

    mid = (s+e)//2
    left_node = make_tree(s, mid, id*2)
    right_node = make_tree(mid+1, e, id*2+1)
    tree[id] = left_node + right_node
    return tree[id]

def query(s, e, id, l, r):
    if s > r or e < l:
        return 0

    if s >= l and e <= r:
        return tree[id]

    mid = (s+e)//2
    left_node = query(s, mid, id*2, l, r)
    right_node = query(mid+1, e, id*2+1, l, r)
    return left_node + right_node

def update(s, e, id, target_id, target_value):
    if target_id < s or target_id > e:
        return

    tree[id] = tree[id] - arr[target_id] + target_value;
    if s == e:
        return

    mid = (s+e)//2
    update(s, mid, id*2, target_id, target_value)
    update(mid+1, e, id*2+1, target_id, target_value)

def solution():
    global q, arr
    make_tree(0, n-1, 1)

    while q > 0:
        l, r, target_id, target_value = map(int, input().split())
        l-=1
        r-=1
        if r < l:
            tmp = r
            r = l
            l = tmp
        print(query(0, n-1, 1, l, r))
        target_id-=1
        update(0, n-1, 1, target_id, target_value)
        arr[target_id] = target_value
        q-=1

if __name__ == '__main__':
    solution()

